Add PipeWire audio driver

j20001970 opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://contributing.godotengine.org/en/latest/engine/guidelines/code_style.html
-->
Closes https://github.com/godotengine/godot-proposals/issues/10723

This PR adds PipeWire audio driver for linuxbsd platform.

## Distro support
This PR expects PipeWire version 1.0.0 or higher. Here is the non-exhaustive list of minimum Linux distribution version required for the driver to work.
<table>
  <tr>
    <th>Name</th>
    <th><code>libpipewire-0.3.so</code> version</th>
  </tr>
  <tr>
    <th>Alpine v3.19</th>
    <th><a href="https://pkgs.alpinelinux.org/package/v3.19/community/x86_64/pipewire">1.0.0</a></th>
  </tr>
  <tr>
    <th>Debian 13</th>
    <th><a href="https://packages.debian.org/trixie/pipewire">1.4.2</a></th>
  </tr>
  <tr>
    <th>Fedora 38</th>
    <th>1.0.3</th>
  </tr>
  <tr>
    <th>openSUSE Leap 15.6</th>
    <th><a href="https://download.opensuse.org/distribution/leap/15.6/repo/oss/x86_64/libpipewire-0_3-0-1.0.5+git36.60deeb2-150600.1.3.x86_64.rpm">1.0.5</a></th>
  </tr>
  <tr>
    <th>Ubuntu 24.04 LTS</th>
    <th><a href="https://packages.ubuntu.com/noble/pipewire">1.0.5</a></th>
  </tr>
</table>

## How to test the driver
The new PipeWire audio driver does not take precedence over existing PulseAudio driver at the moment. You can change `audio/driver/driver` to "PipeWire" in Project Settings, or use `--audio-driver PipeWire` arguments to test the driver.

## Vendored PipeWire headers
This PR introduces new thirdparty headers to enable PipeWire support. A brief list of added PipeWire and SPA headers **as of version 1.0.0** (the list may change if we update the headers in future) is documented to give a better grasp on how these headers are included (no pun intended).

```
Required by "pipewire/pipewire.h":
|───"pipewire/array.h"
|───"pipewire/client.h"
|───"pipewire/conf.h"
|───"pipewire/context.h"
|───"pipewire/device.h"
|───"pipewire/buffers.h"
|───"pipewire/core.h"
|───"pipewire/factory.h"
|───"pipewire/keys.h"
|───"pipewire/log.h"
|───"pipewire/loop.h"
|───"pipewire/link.h"
|───"pipewire/main-loop.h"
|───"pipewire/map.h"
|───"pipewire/mem.h"
|───"pipewire/module.h"
|───"pipewire/node.h"
|───"pipewire/properties.h"
|───"pipewire/proxy.h"
|───"pipewire/permission.h"
|───"pipewire/protocol.h"
|───"pipewire/port.h"
|───"pipewire/stream.h"
|───"pipewire/filter.h"
|───"pipewire/thread-loop.h"
|───"pipewire/data-loop.h"
|───"pipewire/type.h"
|───"pipewire/utils.h"
└───"pipewire/version.h"

SPA headers required by PipeWire:
|───"spa/buffer/buffer.h"
|   └───"spa/buffer/meta.h"
|───"spa/node/command.h"
|───"spa/node/io.h"
|───"spa/node/node.h"
|   └───"spa/node/event.h"
|       └───"spa/pod/event.h"
|───"spa/param/param.h"
|   |───"spa/param/buffers.h"
|   |───"spa/param/port-config.h"
|   |───"spa/param/profile.h"
|   └───"spa/param/route.h"
|───"spa/pod/command.h"
|───"spa/pod/pod.h"
|───"spa/support/log.h"
|───"spa/support/loop.h"
|   └───"spa/support/system.h"
|───"spa/support/plugin.h"
|───"spa/support/thread.h"
|───"spa/utils/defs.h"
|───"spa/utils/dict.h"
|───"spa/utils/hook.h"
|───"spa/utils/list.h"
|───"spa/utils/string.h"
└───"spa/utils/type.h"

Headers required by `AudioDriverPipeWire`:
|───"spa/param/audio/raw.h" <- for SPA_AUDIO_FORMAT_* and SPA_AUDIO_CHANNEL_* enum
|───"spa/param/format.h" <- for SPA_FORMAT_AUDIO_* enum
|───"spa/pod/builder.h" <- for spa pod builder related functions and macros
|   |───"spa/pod/iter.h"
|   └───"spa/pod/vararg.h"
|───"spa/pod/parser.h" <- for spa_pod_parse_object
└───"spa/utils/keys.h"
```

j20001970 committed:
Add PipeWire audio driver

Nintorch commented:
See also https://github.com/godotengine/godot/pull/109362 which may also add PipeWire support
Not to say your PR is worse or anything, just wanted to let you know about this PR too :)
But I'm not sure which one may be better or worse, or just different.

j20001970 commented:
I rebased the commits so that this PR no longer depends on #109500.

j20001970 commented:
9f33ab017ee80cb6581a55fb5f4213957bd97eaa added PipeWire version check and changed headers version to `1.0.0`, in sync with #109500.

j20001970 commented:
f1c07cad65925fa942086cbae37c27d858af8d90 now uses `print_verbose` instead of `ERR_FAIL_NULL_V` macro if `pw_context_connect` failed in  `AudioDriverPipeWire::init`, this is for suppressing errors on audio driver fallback when PipeWire library meets version requirement but the socket is not accessible.

ArchercatNEO commented:
Are we ready for the next round of review?

j20001970 commented:
> Are we ready for the next round of review?

The PR is ready for the next round of review. Should I re-request the review, or wait for the reviewers?

ArchercatNEO commented:
I think re-requesting a review would be good, I asked on RocketChat for a review and that's how the previous round happened. Calinou also said he'd review but I haven't seen him so I think you could ping him too when asking for review

j20001970 commented:
Hi @Calinou, can you help reviewing this PR? Thank you in advance!

Calinou reviewed (commented):
Tested locally on Fedora 43, it mostly works as expected.

Some feedback:

- PipeWire appears to be used natively, according to the [above instructions](https://github.com/godotengine/godot/pull/111056#issuecomment-4074777513):

```bash
$ pw-dump > ~/pw-dump
$ rg -i "platformer" pw-dump
10101:        "application.name": "Platformer 2D",
10168:        "node.name": "Platformer 2D",
10563:        "object.path": "Platformer 2D:output_0",
10565:        "port.alias": "Platformer 2D:output_FL",
10648:        "object.path": "Platformer 2D:output_1",
10650:        "port.alias": "Platformer 2D:output_FR",
```

- Audio input doesn't seem to work in the [Mic Record](https://github.com/godotengine/godot-demo-projects/tree/master/audio/mic_record) demo (nothing is recorded, and the app doesn't appear in the list of recording applications). In comparison, it works in PulseAudio (but not in ALSA).

- PipeWire sounds noticeably quieter than PulseAudio/ALSA. I double-checked all audio drivers are at 100% for the per-app volume in the system settings.

Music from the Platformer 2D demo run through PulseAudio, ALSA, then PipeWire:

[pulseaudio_alsa_pipewire.mp3](https://github.com/user-attachments/files/26729828/pulseaudio_alsa_pipewire.mp3)

Waveform of the above audio normalized in Audacity for easier comparison (from left to right: PulseAudio, ALSA, PipeWire):

<img width="916" height="306" alt="Image" src="https://github.com/user-attachments/assets/6b586cfe-db38-4101-a0b9-af16f79aa8c5" />

- I haven't done latency testing since I don't have the hardware for it currently, but it would be good to investigate this in the future.

j20001970 commented:
> Tested locally on Fedora 43, it mostly works as expected.
> 
> Some feedback:
> 
>     * PipeWire appears to be used natively, according to the [above instructions](https://github.com/godotengine/godot/pull/111056#issuecomment-4074777513):
> 
> 
> ```shell
> $ pw-dump > ~/pw-dump
> $ rg -i "platformer" pw-dump
> 10101:        "application.name": "Platformer 2D",
> 10168:        "node.name": "Platformer 2D",
> 10563:        "object.path": "Platformer 2D:output_0",
> 10565:        "port.alias": "Platformer 2D:output_FL",
> 10648:        "object.path": "Platformer 2D:output_1",
> 10650:        "port.alias": "Platformer 2D:output_FR",
> ```
> 
>     * Audio input doesn't seem to work in the [Mic Record](https://github.com/godotengine/godot-demo-projects/tree/master/audio/mic_record) demo (nothing is recorded, and the app doesn't appear in the list of recording applications). In comparison, it works in PulseAudio (but not in ALSA).
> 
>     * PipeWire sounds noticeably quieter than PulseAudio/ALSA. I double-checked all audio drivers are at 100% for the per-app volume in the system settings.
> 
> 
> Music from the Platformer 2D demo run through PulseAudio, ALSA, then PipeWire:
> 
> [pulseaudio_alsa_pipewire.mp3](https://github.com/user-attachments/files/26729828/pulseaudio_alsa_pipewire.mp3)
> 
> Waveform of the above audio normalized in Audacity for easier comparison (from left to right: PulseAudio, ALSA, PipeWire):
> <img alt="Image" width="916" height="306" src="https://private-user-images.githubusercontent.com/180032/578273928-6b586cfe-db38-4101-a0b9-af16f79aa8c5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzYyMjUzNzIsIm5iZiI6MTc3NjIyNTA3MiwicGF0aCI6Ii8xODAwMzIvNTc4MjczOTI4LTZiNTg2Y2ZlLWRiMzgtNDEwMS1hMGI5LWFmMTZmNzlhYThjNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDQxNVQwMzUxMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zODk3NmEyMzE4NmFjZDc0OGZlOTgzMTRjMmI4OWYxMGU5ZjU4NjlkYTAxNjlkODRkNWZlYmViYTYwM2NhMTY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.MyWV-pccsZ8-vmjDLIUIznADxAmz_zbCswYhDN-lTMs">
> 
>     * I haven't done latency testing since I don't have the hardware for it currently, but it would be good to investigate this in the future.

Thanks for the review, the PR has been updated to 506940fb6ec61dd2907401a659778d036423bcb3 with the following change:
- Audio position [for input stream](https://github.com/godotengine/godot/blob/506940fb6ec61dd2907401a659778d036423bcb3/drivers/pipewire/audio_driver_pipewire.cpp#L557) has been added (the primary reason why input stream didn't work as expected).
- When initializing input stream, [`set_input_device("Default")`](https://github.com/godotengine/godot/blob/506940fb6ec61dd2907401a659778d036423bcb3/drivers/pipewire/audio_driver_pipewire.cpp#L339) is called to update input channel count.
- Init / Start logic for output stream has been updated to remove the redundant `active` SafeFlag from the driver.

For the volume issue across different drivers, I can't seem to reproduce the same problem. The default per-app volume can be different when switching to different driver, but the waveform magnitude looks like the same when comparing with PulseAudio(first) and PipeWire(second) at 100% volume. More testing might be needed for this part.
<img width="1023" height="297" src="https://github.com/user-attachments/assets/c03f5637-fee8-48c3-869f-83b46eb0c05d" />

Calinou reviewed (approved):
Audio input works on my end now. Code looks good to me.

I can still reproduce the different volume on PipeWire, but as per the above comment, it's likely system-specific. I consider this PR good on my end, but I still would like it to get testing on more configurations (especially 3.1/5.1/7.1 for people who have access to such a setup).

Note that this PR uses C++20 features (designated initializers), but we considered this to be acceptable in Wayland code as it's only compiled on Linux/*BSD. Similarly, the PipeWire audio driver is Linux/*BSD-only code.

majenkotech commented:
I'm currently failing to get this PR working with a multichannel setup. I have an 8-channel audio sink in PipeWire, and I'm selecting that sink with `AudioServer.set_output_device("8_channel_sink")`. Godot is reporting 3 (7.1 surround) for `AudioServer.get_speaker_mode()`.

However in `qpwgraph` it seems that Godot is only creating a 2-channel source for the game. Anything sent to Surround or Center is just getting lost as there's no output from Godot to connect for those speakers.

`spa_pod_builder_add_object()` is getting `8` for the number of channels in the pipewire object to construct, but doesn't seem to be actually creating an 8-channel object, only a 2-channel one.

majenkotech commented:
Ok, don't worry. It is, as you can imagine, a PEBCAK. The sink's ports have to be named specifically by the surround sound standard, even if they are not, strictly speaking, surround sound sinks.

majenkotech commented:
Ok, this one I feel sure is a bug. Probably some order-of-execution thing.

If you route your audio manually to a multi-channel device with `AudioServer.set_output_device()` all is groovy. If you set the multi-channel device as the default audio and don't manually route the audio all is groovy. However if the default is not the multi-channel, and you don't manually route it in your code, but instead use KDE's "Applications" feature in the volume control to tie your game to a specific multi-channel device, it's not so groovy. 8 channels are created. 8 channels are connected. But only 2 channels are actually used.

Supposition: it's first connecting to the default, setting the core to generate audio as 2 channel (down-mixing surround into stereo), then the OS is reconfiguring the audio interfaces to 8 channel. But the core is still working in 2-channel down-mixed mode. Or something.

Interestingly, starting the game connected to a 2 channel device, then after a delay redirecting it to the multi-channel in code doesn't show this problem, so it's something to do with the external signal to say "It's got to use this device". Not knowing how any of that works, I can't really say any more...

majenkotech commented:
So some more digging.  PipeWire sends a `param_changed` event that needs to be handled. That sends an updated POD (I think) inside of which is such things as the new channel count and arrangement, and other things related to the audio format (bit depth, sample rate, etc) all of which really need to be handled by the driver to reconfigure the core audio system as needed.

There is also an acknowledge signal that needs to be sent back (from what I have read) by calling `pw_stream_update_params()` - though I have only seen that in passing and can't comment on its validity.

POD has a pod/parser.h with `spa_pod_parse_object()` in it which is supposed to be used to get the updated settings out of the incoming POD, but that needs extra things with it to get it working and I am not familiar enough with scons and the PipeWire driver & linuxbsd file structure to even attempt getting that going.

Needless to say, though, I would label this PR as incomplete until the `param_changed` signal is properly handled by the driver, since that is a pretty fundamental part of the PipeWire protocol.

j20001970 commented:
> So some more digging. PipeWire sends a `param_changed` event that needs to be handled. That sends an updated POD (I think) inside of which is such things as the new channel count and arrangement, and other things related to the audio format (bit depth, sample rate, etc) all of which really need to be handled by the driver to reconfigure the core audio system as needed.
> 
> There is also an acknowledge signal that needs to be sent back (from what I have read) by calling `pw_stream_update_params()` - though I have only seen that in passing and can't comment on its validity.
> 
> POD has a pod/parser.h with `spa_pod_parse_object()` in it which is supposed to be used to get the updated settings out of the incoming POD, but that needs extra things with it to get it working and I am not familiar enough with scons and the PipeWire driver & linuxbsd file structure to even attempt getting that going.
> 
> Needless to say, though, I would label this PR as incomplete until the `param_changed` signal is properly handled by the driver, since that is a pretty fundamental part of the PipeWire protocol.

Thanks for the detailed testing. You are right about that the driver does not keep track of current sink when changing audio device externally (e.g. from volume control panel). I tried to address this issue a while back but to no avail, but I'll see if `param_changed` could provide formats for newly selected audio device.

j20001970 commented:
Did some more testing on **how to detect output device change of the stream when specified externally**, so it's time for report. Tested on PipeWire 1.6.4, WirePlumber 0.5.14, KDE Plasma 6.6.4

- In the case of `param_changed` stream event, **when connecting the stream from the driver**, the event would usually report with id **2**(`SPA_PARAM_Props`), **4**(`SPA_PARAM_Format`), **15**(`SPA_PARAM_Latency`) and **17**(`SPA_PARAM_Tag`).
Among these param types, `SPA_PARAM_Format` can be useful for determining channel count. But **when the output device of the stream is changed from volume control panel**, ~~only `SPA_PARAM_Latency` is reported, which doesn't reveal any info about newly selected output device.~~
Changing the default sink does not count, as `AudioDriverPipeWire::on_metadata_property` will reconnect the stream if current device is `Default`.
**EDIT**: After tinkering a bit more, `SPA_PARAM_Format` will also report but not as reliably as I thought, I guess this need more investigation.
- Other stream events are also attempted, `control_info`, `command` events do not emit when changing output device from volume control panel, while `io_changed` event will emit with id **7**(`SPA_IO_Position`), which relevant struct `spa_io_position` is about raw clock times.
- When changing the stream output device externally, **stream properties does not reflect newly selected target object**. So checking for stream `PW_KEY_TARGET_OBJECT` property in every `process` event won't work.
- `AudioDriverPipeWire::on_metadata_property` metadata event seems to be able to detect change in stream output target for **all** streams (this includes other non-Godot processes/apps) with key-value pair
{`target.node`: output sink id}, but I haven't found a way to associate the subject id with the driver.

Concluding what I've found currently, I think ~~the `AudioDriverPipeWire::on_metadata_property` approach is the most promising way to achieve detecting externally changed output device of the stream if such functionality is desired~~. And `param_changed` event might be a more ideal way to update stream channel count as part of format negotiation when connecting the stream.

j20001970 commented:
After figuring out how to get stream channel count reliably, cb22c4c7fcc2b43245ac1749589ef9efb242b6e8 has made quite some change to improve multi-channel handling:

- Stream channel count are now updated from `param_changed` event instead of using node port count, and `SPA_FORMAT_AUDIO_channels` is removed from stream param to get actual stream channel count in `param_changed` event.
This change means that `AudioServer.get_speaker_mode()` will based on channels actually connected to the sink node, instead of total ports of the connected node.
- Since channel count is now updated from `param_changed` stream event, `info` event for sink/source nodes are also removed.
- Tracking for current default sink/source node in order to get channel count in advance is no longer needed, so metadata `property` event is removed as well.
- Added `state_changed` event for both output and input stream, could be useful for future debugging.
- Improved uneven channels handling for both output stream (e.g. playing 4-channel samples to 3-channel sink, or connecting to sink node with more than 8 channels) and input stream (in the case that the source node has more than 2 channels).

j20001970 commented:
35d8077923b6f17ffa4b553efaa6c3d229bb23e5 added mono playback handling (by averaging stereo channels), and the driver now uses `audio/driver/output_latency` from project settings to configure buffer size for output stream.

Chlipouni commented:
Thanks a lot for this PR.
I'm currently using it in my project under Ubuntu Studio 26.04.
It works very well and allow more consistent buffer size settings for audio low latency.

but sometimes, I have the following Godot crash :
```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.7.beta.custom_build
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
Load address: 7d806d800000
[1] 7d806d845cb0 (main+45cb0) - /usr/lib/x86_64-linux-gnu/libc.so.6(+0x45cb0) [0x7d806d845cb0]
[2] 6434ea876d7e (godot.linuxbsd.editor.x86_64+4a00d7e) - std::ctype<char>::do_widen(char) const à ??:?
[3] 6434ea877cc4 (godot.linuxbsd.editor.x86_64+4a01cc4) - std::ctype<char>::do_widen(char) const à ??:?
[4] 6434e7e26894 (godot.linuxbsd.editor.x86_64+1fb0894) - std::ctype<char>::do_widen(char) const à ??:?
[5] 7d806838cad8 (libpipewire-0.3.so.0+a0ad8) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0xa0ad8) [0x7d806838cad8]
[6] 7d8003f00113 (libspa-audioconvert.so+1c113) - /usr/lib/x86_64-linux-gnu/spa-0.2/audioconvert/libspa-audioconvert.so(+0x1c113) [0x7d8003f00113]
[7] 7d8068362547 (libpipewire-0.3.so.0+76547) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0x76547) [0x7d8068362547]
[8] 7d80683629d6 (libpipewire-0.3.so.0+769d6) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0x769d6) [0x7d80683629d6]
[9] 7d80682c6a96 (libspa-support.so+9a96) - /usr/lib/x86_64-linux-gnu/spa-0.2/support/libspa-support.so(+0x9a96) [0x7d80682c6a96]
[10] 7d8068329903 (libpipewire-0.3.so.0+3d903) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0x3d903) [0x7d8068329903]
[11] 7d806d8a40da (main+a40da) - /usr/lib/x86_64-linux-gnu/libc.so.6(+0xa40da) [0x7d806d8a40da]
[12] 7d806d9377ac (main+1377ac) - /usr/lib/x86_64-linux-gnu/libc.so.6(+0x1377ac) [0x7d806d9377ac]
-- END OF C++ BACKTRACE --
================================================================
```

This crash occurs when Raysession and Guitarix are already open when I launch my Godot Project.
The crash occurs randomly as I am able to launch my Godot Project the second or third try.

Ask me if you need more trace files.

j20001970 commented:
Hi @Chlipouni, can you try reproducing the crash with `PIPEWIRE_DEBUG=T` environment variable?

A log file could be useful by redirecting the output like this:
`PIPEWIRE_DEBUG=T ./bin/godot.linuxbsd.editor.x86_64 &> trace.log`

Chlipouni commented:
@j20001970, find the trace in attached file :
[trace.log](https://github.com/user-attachments/files/28434546/trace.log)

j20001970 commented:
> @j20001970, find the trace in attached file : [trace.log](https://github.com/user-attachments/files/28434546/trace.log)

The audio driver setup process seems working as expected from the look of PipeWire trace log, but crashed soon after a few `pw_thread_loop_lock/unlock`s. I guess the crash come from output stream `process` event, but I'm not yet certain which part caused that.

@Chlipouni can you try reproducing the crash again, with `debug_symbols=yes` this time when building Godot? The Godot crash backtrace alone should be good for locating the problem, thanks!

Chlipouni commented:
@j20001970, find the trace in attached file : [trace.log](https://github.com/user-attachments/files/28444589/trace.log)

The backtrace :
```
WARNING: Image format RGB8 not supported by hardware, converting to RGBA8.
     at: _validate_texture_format (servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:2829)
     GDScript backtrace (most recent call first):
         [0] @implicit_new (res://scripts/global.gd:49)
ERROR: Parameter "input_stream" is null.
   at: input_stop (drivers/pipewire/audio_driver_pipewire.cpp:528)

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.7.beta.custom_build
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
Load address: 741c3d400000

WARNING: Image format RGB8 not supported by hardware, converting to RGBA8.
     at: _validate_texture_format (servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:2829)
     GDScript backtrace (most recent call first):
         [0] @implicit_new (res://scripts/global.gd:49)
[1] 741c3d445cb0 (main+45cb0) - /usr/lib/x86_64-linux-gnu/libc.so.6(+0x45cb0) [0x741c3d445cb0]
[2] 62acb72b5d7e (godot.linuxbsd.editor.x86_64+4a00d7e) - AudioServer::_mix_step() à /home/ogc/VS Code/godot-pw/servers/audio/audio_server.cpp:578 (discriminator 6)
[3] 62acb72b6cc4 (godot.linuxbsd.editor.x86_64+4a01cc4) - AudioServer::_driver_process(int, int*) à /home/ogc/VS Code/godot-pw/servers/audio/audio_server.cpp:292
[4] 62acb4865894 (godot.linuxbsd.editor.x86_64+1fb0894) - AudioDriverPipeWire::on_output_stream_process(void*) à /home/ogc/VS Code/godot-pw/drivers/pipewire/audio_driver_pipewire.cpp:199 (discriminator 1)
[5] 741c340dcad8 (libpipewire-0.3.so.0+a0ad8) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0xa0ad8) [0x741c340dcad8]
[6] 741bf4032113 (libspa-audioconvert.so+1c113) - /usr/lib/x86_64-linux-gnu/spa-0.2/audioconvert/libspa-audioconvert.so(+0x1c113) [0x741bf4032113]
[7] 741c340b2547 (libpipewire-0.3.so.0+76547) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0x76547) [0x741c340b2547]
[8] 741c340b29d6 (libpipewire-0.3.so.0+769d6) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0x769d6) [0x741c340b29d6]
[9] 741c34016a96 (libspa-support.so+9a96) - /usr/lib/x86_64-linux-gnu/spa-0.2/support/libspa-support.so(+0x9a96) [0x741c34016a96]
[10] 741c34079903 (libpipewire-0.3.so.0+3d903) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0x3d903) [0x741c34079903]
[11] 741c3d4a40da (main+a40da) - /usr/lib/x86_64-linux-gnu/libc.so.6(+0xa40da) [0x741c3d4a40da]
[12] 741c3d5377ac (main+1377ac) - /usr/lib/x86_64-linux-gnu/libc.so.6(+0x1377ac) [0x741c3d5377ac]
-- END OF C++ BACKTRACE --
================================================================
```

hpvb reviewed (changes requested):
This needs some discussion. 

This is a temporary nak to avoid merge for the moment.

j20001970 commented:
> @j20001970, find the trace in attached file : [trace.log](https://github.com/user-attachments/files/28444589/trace.log)
> 
> The backtrace :
> 
> ```
> WARNING: Image format RGB8 not supported by hardware, converting to RGBA8.
>      at: _validate_texture_format (servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:2829)
>      GDScript backtrace (most recent call first):
>          [0] @implicit_new (res://scripts/global.gd:49)
> ERROR: Parameter "input_stream" is null.
>    at: input_stop (drivers/pipewire/audio_driver_pipewire.cpp:528)
> 
> ================================================================
> handle_crash: Program crashed with signal 11
> Engine version: Godot Engine v4.7.beta.custom_build
> Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
> Load address: 741c3d400000
> 
> WARNING: Image format RGB8 not supported by hardware, converting to RGBA8.
>      at: _validate_texture_format (servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:2829)
>      GDScript backtrace (most recent call first):
>          [0] @implicit_new (res://scripts/global.gd:49)
> [1] 741c3d445cb0 (main+45cb0) - /usr/lib/x86_64-linux-gnu/libc.so.6(+0x45cb0) [0x741c3d445cb0]
> [2] 62acb72b5d7e (godot.linuxbsd.editor.x86_64+4a00d7e) - AudioServer::_mix_step() à /home/ogc/VS Code/godot-pw/servers/audio/audio_server.cpp:578 (discriminator 6)
> [3] 62acb72b6cc4 (godot.linuxbsd.editor.x86_64+4a01cc4) - AudioServer::_driver_process(int, int*) à /home/ogc/VS Code/godot-pw/servers/audio/audio_server.cpp:292
> [4] 62acb4865894 (godot.linuxbsd.editor.x86_64+1fb0894) - AudioDriverPipeWire::on_output_stream_process(void*) à /home/ogc/VS Code/godot-pw/drivers/pipewire/audio_driver_pipewire.cpp:199 (discriminator 1)
> [5] 741c340dcad8 (libpipewire-0.3.so.0+a0ad8) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0xa0ad8) [0x741c340dcad8]
> [6] 741bf4032113 (libspa-audioconvert.so+1c113) - /usr/lib/x86_64-linux-gnu/spa-0.2/audioconvert/libspa-audioconvert.so(+0x1c113) [0x741bf4032113]
> [7] 741c340b2547 (libpipewire-0.3.so.0+76547) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0x76547) [0x741c340b2547]
> [8] 741c340b29d6 (libpipewire-0.3.so.0+769d6) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0x769d6) [0x741c340b29d6]
> [9] 741c34016a96 (libspa-support.so+9a96) - /usr/lib/x86_64-linux-gnu/spa-0.2/support/libspa-support.so(+0x9a96) [0x741c34016a96]
> [10] 741c34079903 (libpipewire-0.3.so.0+3d903) - /usr/lib/x86_64-linux-gnu/libpipewire-0.3.so.0(+0x3d903) [0x741c34079903]
> [11] 741c3d4a40da (main+a40da) - /usr/lib/x86_64-linux-gnu/libc.so.6(+0xa40da) [0x741c3d4a40da]
> [12] 741c3d5377ac (main+1377ac) - /usr/lib/x86_64-linux-gnu/libc.so.6(+0x1377ac) [0x741c3d5377ac]
> -- END OF C++ BACKTRACE --
> ================================================================
> ```

Hi @Chlipouni, is it possible to reproduce the crash with empty project? I'm currently trying to reproduce the problem from my end, is there any specific steps on configuring Raysession and Guitarix?

Chlipouni commented:
@j20001970,
I created a new project with only a few scenes and scripts to manage audio input and output.
I can't reproduce the crash with your Godot fork.

So, it seems that the crash is related to either :
- my project
- my own custom build of Godot

I am very sorry for this inconvenience.

j20001970 commented:
> @j20001970, I created a new project with only a few scenes and scripts to manage audio input and output. I can't reproduce the crash with your Godot fork.
> 
> So, it seems that the crash is related to either :
> 
>     * my project
> 
>     * my own custom build of Godot
> 
> 
> I am very sorry for this inconvenience.

No wonder the source code line numbers seems a bit off in the backtrace...

In that case, is it possible to test whether this PR (commit 900c044655217a3a5a277acee2ac98696c77c1e5) work on your project?

Chlipouni commented:
@j20001970,
My issue seems to be related to the Godot RTMidi extension I'm using in my project to manage MIDI message output.
Sometimes, when I open my project and RaySession, audio is working fine, but the link between RTMidi and Guitarix MIDI input is not set. When I try to link again the MIDI ports, it randomly breaks the audio.
I will do more tests without the RTMidi extension...

ArchercatNEO commented:
I was testing the PW CPU usage with the Godot profiler and was getting anomalous results.

PulseAudio
<img width="996" height="108" alt="PulseAudioCpuUsage" src="https://github.com/user-attachments/assets/d9ef32f3-0333-4afa-bb61-1b32b2b9aad2" />

PipeWire
<img width="414" height="209" alt="PipeWireCpuUsage" src="https://github.com/user-attachments/assets/d41fcb2d-6fab-4f4a-954e-3be0427b77f4" />

That PW was taking 0ms seemed very suspicious so I looked a bit more at `start_counting_ticks`.

The PA driver counts ticks when copying from the Audio server to local buffers (which the PW driver also does) but also measures time spent iterating the main loop (which the PW driver doesn't do).
https://github.com/godotengine/godot/blob/3badc0ee23caf1e5269fe90f8ee17e09d4d2b682/drivers/pulseaudio/audio_driver_pulseaudio.cpp#L447-L453

Shouldn't we be using plain loops (https://docs.pipewire.org/group__pw__loop.html) that we iterate ourselves  to measure the CPU usage of the PW loop like we do for the PA loop?

j20001970 commented:
> I was testing the PW CPU usage with the Godot profiler and was getting anomalous results.
> 
> PulseAudio <img alt="PulseAudioCpuUsage" width="996" height="108" src="https://private-user-images.githubusercontent.com/108980279/602135538-d9ef32f3-0333-4afa-bb61-1b32b2b9aad2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODA0ODM5OTEsIm5iZiI6MTc4MDQ4MzY5MSwicGF0aCI6Ii8xMDg5ODAyNzkvNjAyMTM1NTM4LWQ5ZWYzMmYzLTAzMzMtNGFmYS1iYjYxLTFiMzJiMmI5YWFkMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNjAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDYwM1QxMDQ4MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYWIwY2YwZjYyM2FkOWEzNmFkNTRjMmU5M2FiODFhNDVmOTg3NDJmNmZkMTgyNWZkNTBkY2Q5Yzg2MzMzN2FjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.nOemimJa0zBeoMs2n0hmC1vGdpzcJHHSzyctI6upFVo">
> 
> PipeWire <img alt="PipeWireCpuUsage" width="414" height="209" src="https://private-user-images.githubusercontent.com/108980279/602135790-d41fcb2d-6fab-4f4a-954e-3be0427b77f4.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODA0ODM5OTEsIm5iZiI6MTc4MDQ4MzY5MSwicGF0aCI6Ii8xMDg5ODAyNzkvNjAyMTM1NzkwLWQ0MWZjYjJkLTZmYWItNGY0YS05NTRlLTNiZTA0MjdiNzdmNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNjAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDYwM1QxMDQ4MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05N2M2YzY0ODFkMjMwYjQ0MzJkYjk3YWQ1MzNjNjNiZjEwMTIzZThhYWNkY2JhN2M1NzNmYTNmZGJlMzg4MmI5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.ZPxDDckWMS75SCPiM6_5KEO4t2hVfimV5f_pKv-Jcy4">
> 
> That PW was taking 0ms seemed very suspicious so I looked a bit more at `start_counting_ticks`.
> 
> The PA driver counts ticks when copying from the Audio server to local buffers (which the PW driver also does) but also measures time spent iterating the main loop (which the PW driver doesn't do).
> 
> https://github.com/godotengine/godot/blob/3badc0ee23caf1e5269fe90f8ee17e09d4d2b682/drivers/pulseaudio/audio_driver_pulseaudio.cpp#L447-L453
> 
> Shouldn't we be using plain loops (https://docs.pipewire.org/group__pw__loop.html) that we iterate ourselves to measure the CPU usage of the PW loop like we do for the PA loop?

I think the audio driver time in the debugger `Monitors` tab comes from `AudioDriver::get_latency`, which is not yet implemented in PipeWire driver since I was not sure on how to measure that.

I'll look into it again to see how to implement this method.

j20001970 commented:
The PR has been updated to d2b7d50046283e32c310a021ecbdb57083184d62 with the following changes:

- `AudioDriverPipeWire::get_latency` is now implemented. Latency calculation is based on PipeWire documentation [pw_time](https://docs.pipewire.org/structpw__time.html#details):

> For an audio playback stream, if you were to queue a buffer, the total delay in milliseconds for the first sample in the newly queued buffer to be played by the hardware can be calculated as:
> ```c
> (pw_time.buffered * 1000 / stream.samplerate) +
>   (pw_time.queued * 1000 / app.rate) +
>    ((pw_time.delay - elapsed) * 1000 * pw_time.rate.num / pw_time.rate.denom)
> ```

- `AudioDriverPipeWire::input_stop` error is silenced when input stream is destroyed before the call.
- The driver now use `PW_KEY_NODE_LATENCY` stream property to configure output stream buffer size from `audio/driver/output_latency`. The previous `SPA_PARAM_BUFFERS_size` param approach doesn't seem to have effect on `AudioDriverPipeWire::get_latency`.

ArchercatNEO commented:
Tested again, and now we do get some numbers:
<img width="990" height="199" alt="pipewire-latency" src="https://github.com/user-attachments/assets/1d62a84d-d7c8-44d2-a151-c51feae3258e" />

This was tested with the `audio-effects` sample project and me spamming as many buttons as I could, and indeed we got latency improvements. The minimum pulse latency was higher than the maximum pipewire latency, but the main point is whether this is relevant for most people.

I remember two of the reasons listed for this PR are lower latency (which we can put numbers on now) and that ALSA didn't support input. From the dev chat, there's a sentiment that fixing ALSA input would be simpler than implementing a pipewire driver. Other than these reasons, and that pipewire is useful for screen capture anyway (which is a separate PR), are there more benefits to having this pipewire audio driver?

j20001970 commented:
> I remember two of the reasons listed for this PR are lower latency (which we can put numbers on now) and that ALSA didn't support input. From the dev chat, there's a sentiment that fixing ALSA input would be simpler than implementing a pipewire driver. Other than these reasons, and that pipewire is useful for screen capture anyway (which is a separate PR), are there more benefits to having this pipewire audio driver?

I suppose achieving lower latency is the primary and immediate benefit of including a native PipeWire driver, along with theoretically lower CPU usage as the result of skipping `pipewire-pulse`.

In long-term, adopting PipeWire API in audio driver enables the flexibility of what PipeWire provides (which is already the default audio server on several Linux distributions) as Godot audio subsystem continues to evolve. PulseAudio remains the default driver for maximum compatibility, but including PipeWire as a option gives developers and users a opportunity to test the driver for performance and stability before PulseAudio is eventually phased out.

Note that ALSA driver currently only support stereo output, and it seems that only "Default" device is listed in `device_changer` demo project (this might be system-specific though). So more work might be needed for ALSA driver in addition to fixing input to have the same features as PulseAudio/PipeWire driver.

Chlipouni commented:
@j20001970, FYI if I use your fork for Pipewire audio driver I don't have the Godot crash issue anymore.
It seems that the crashes are due to my own modifications in the code (my own custom build).
